docs(coding-guidelines): Add request response correlation

mxinden opened the pull request:
## Description

Allow correlating asynchronous responses to their requests.

<!-- Please write a summary of your changes and why you made them.-->
<!-- This section will appear as the commit message after merging. Please craft it accordingly. -->

## Notes

<!-- Any notes or remarks you'd like to make about the PR. -->

## Links to any relevant issues

Came up in discussions around https://github.com/libp2p/rust-libp2p/issues/2824.

## Open Questions

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [ ] ~~A changelog entry has been made in the appropriate crates~~

mxinden committed:
docs(coding-guidelines): Add request response correlation

Allow correlating asynchronous responses to their requests.

thomaseizinger reviewed (changes requested):
Thanks for writing this up!

I think we should settle on one approach on how we do this in the codebase, otherwise it is not really a guideline.

My preference would be client-side generated IDs for commands:

They are a special case of "user data" just not being configurable, thus they can carry the same usecases but without the downsides of type parameters which make the code complex.

I am not sure about methods. For https://github.com/libp2p/rust-libp2p/pull/3292, we would need client-side generated IDs but we can't do that if we have a more low-level component that hands out IDs.

elenaf9 commented on docs/coding-guidelines.md:
```diff
@@ -299,3 +300,55 @@ stack potentially unboundedly. Instead use iteration e.g. via `loop` or `for`.
+## Allow Correlating Asynchronous Responses to Their Requests
+
+In an asynchronous context, it is important to enable users to determine the correlation between a
+response and a previous request. For example, if a user requests two new connections to the same
+peer, they should be able to match each new connection to the corresponding previous connection
+request without having to guess.
```
I agree with the general guideline, but I am not sure if I agree with this example. Connections can still be differentiated by the `ConnectedPoint`. And if that's also the same, why does it matter what connection was initiated by what request?

mxinden replied:
Say that I have two `NetworkBehaviour`s `A` and `B`. Both dial a peer concurrently. `A` requests a dial to the peer with a set of addresses. `B` requests a dial to the peer with a set of addresses. `A` keeps some state during the time of the dial. `A` is notified that a connection request to the peer failed. Should `A` delete the state? Maybe this was the dial of `B` with the wrong set of addresses.

elenaf9 commented on docs/coding-guidelines.md:
````diff
@@ -299,3 +300,55 @@ stack potentially unboundedly. Instead use iteration e.g. via `loop` or `for`.
+``` rust
+fn my_method() -> impl Future<Output = Response> {
+  // ...
+}
````
Are we doing this anywhere in rust-libp2p, apart from `Transport::dial`?

mxinden replied:
`Tranport::dial` is the only one that comes to my mind. That said, I would like to experiment with it on the API surface of `NetworkBehaviour` implementations. E.g. a `NotBitswap` `NetworkBehaviour` implementation providing a `request_cid` method that returns a `Future`.

I don't have a strong preference, nor a rule for either of the two (return ID or return Future). Thus I suggest keeping both options open. Objections?

elenaf9 commented on docs/coding-guidelines.md:
```diff
@@ -299,3 +300,55 @@ stack potentially unboundedly. Instead use iteration e.g. via `loop` or `for`.
+When accepting a **command** that eventually results in a response through an event, either require
+that command to contain a unique ID, which is later on contained in the asynchronous response event,
+or allow the user to provide an instance of a custom type which again is later on returned in the
+response event.
```
Could we briefly elaborate where such commands are used rust-libp2p? I would consider the events from `NetworkBehaviour` -> `Handler` as such commands, or did you have something different in mind?

mxinden replied:
694a09ce adds the example of `NetworkBehaviourAction` from `NetworkBehaviour` to `Swarm`. `NetworkBehaviour` -> `Handler` is an option as well, though in my eyes the less ideal example as it has the indirection through `Swarm`.

elenaf9 commented on docs/coding-guidelines.md:
````diff
@@ -299,3 +300,55 @@ stack potentially unboundedly. Instead use iteration e.g. via `loop` or `for`.
+fn my_method() -> Id {
+  // ...
+}
+```
````
Compared to the `Command` below the ID here is not user generated but instead handed out to the user. When would it be recommended to use user-generated IDs and when should they be handed out?

mxinden replied:
I don't have a good rule here, nor an opinion. Thus far I was simply deciding case-by-case.

elenaf9 reviewed (commented):
Thanks for writing this up @mxinden. 
Right now, this describes different approaches to solve more or less the same thing. I think it would be useful if there would be some additional notes on when it makes sense to use what.

mxinden committed:
Remove user data command option

mxinden committed:
Add example for Command from NetworkBehaviour to Swarm

Note that this is only really valid once we have https://github.com/libp2p/rust-libp2p/pull/3327 and
the corresponding patch in `libp2p-swarm` `DialOpts`.

mxinden committed:
Return leftover brackets

thomaseizinger commented on docs/coding-guidelines.md:
````diff
@@ -299,3 +300,42 @@ stack potentially unboundedly. Instead use iteration e.g. via `loop` or `for`.
+fn my_method() -> Id {
+  // ...
+}
+```
````
Would it make sense to remove this and be consistent with commands where the caller always specifies the ID? Only `Transport::dial` currently behaves like stated here and we already said we wanted to fix this! :)

I'd be in favor of removing this from the guidelines (or rather, say that we want to accept the ID) and opening an issue to fix `Transport::dial`.

mxinden replied:
> with commands where the caller always specifies the ID?

I have to give this more thought. In other words, I am not sure this is the way to go in all situations.

Do you feel strongly about removing this @thomaseizinger? I would prefer proceeding here. I see this pull request as a net-positive at this point.

thomaseizinger replied:
I think it makes sense for anything that is public API in our library. Internally, where we control all usages of an ID, it doesn't matter as much.

The reason for that is that the majority of our public API is command-based. Anything command-based needs to create the ID ahead of time. Any component underneath that creates its own IDs causes friction here because it doesn't compose well with commands.

thomaseizinger replied:
I am a bit torn. There is certainly value in documenting this but if we document something, I'd prefer if we can back the written words 100%, e.g. have strong convictions that this is the way to go. Everything else is just confusing to the reader.

thomaseizinger replied:
Would removing this section be a middle ground? i.e. only document that commands should come with a user-generated ID and not talk about functions at all?

mxinden replied:
Done with 12e61b63.

mxinden committed:
Remove mention of function

mergify[bot] committed:
Merge branch 'master' into guidelines-response-correlation
